homeassistant/components/greeneye_monitor/sensor.py

```diff
@@ -46,55 +46,58 @@
 COUNTER_ICON = "mdi:counter"
 
 
-async def async_setup_platform(
+async def async_setup_entry(
     hass: HomeAssistant,
-    config: Config,
+    config_entry: ConfigEntry,
     async_add_entities: AddEntitiesCallback,
-    discovery_info: DiscoveryInfoType,
 ) -> None:
-    """Set up a single GEM temperature sensor."""
+    """Set up GEM sensors from the config entry."""
     entities: list[GEMSensor] = []
-    for sensor in discovery_info[CONF_SENSORS]:
-        sensor_type = sensor[CONF_SENSOR_TYPE]
-        if sensor_type == SENSOR_TYPE_CURRENT:
+    for monitor_config in config_entry.options[CONF_MONITORS]:
+        serial_number = monitor_config[CONF_SERIAL_NUMBER]
+
+        channels_options = monitor_config[CONF_CHANNELS]
+        for sensor in channels_options:
             entities.append(
                 CurrentSensor(
-                    sensor[CONF_MONITOR_SERIAL_NUMBER],
+                    serial_number,
                     sensor[CONF_NUMBER],
                     sensor[CONF_NAME],
                     sensor[CONF_NET_METERING],
                 )
             )
-        elif sensor_type == SENSOR_TYPE_PULSE_COUNTER:
+
+        pulse_counters_options = monitor_config[CONF_PULSE_COUNTERS]
+        for sensor in pulse_counters_options:
             entities.append(
                 PulseCounter(
-                    sensor[CONF_MONITOR_SERIAL_NUMBER],
+                    serial_number,
                     sensor[CONF_NUMBER],
                     sensor[CONF_NAME],
                     sensor[CONF_COUNTED_QUANTITY],
                     sensor[CONF_TIME_UNIT],
                     sensor[CONF_COUNTED_QUANTITY_PER_PULSE],
                 )
             )
-        elif sensor_type == SENSOR_TYPE_TEMPERATURE:
+
+        temperature_sensors_options = monitor_config[CONF_TEMPERATURE_SENSORS]
+        for sensor in temperature_sensors_options[CONF_SENSORS]:
             entities.append(
                 TemperatureSensor(
-                    sensor[CONF_MONITOR_SERIAL_NUMBER],
+                    serial_number,
                     sensor[CONF_NUMBER],
                     sensor[CONF_NAME],
-                    sensor[CONF_TEMPERATURE_UNIT],
+                    temperature_sensors_options[CONF_TEMPERATURE_UNIT],
                 )
             )
-        elif sensor_type == SENSOR_TYPE_VOLTAGE:
+
+        voltage_sensors_options = monitor_config[CONF_VOLTAGE_SENSORS]
+        for sensor in voltage_sensors_options:
             entities.append(
-                VoltageSensor(
-                    sensor[CONF_MONITOR_SERIAL_NUMBER],
-                    sensor[CONF_NUMBER],
-                    sensor[CONF_NAME],
-                )
+                VoltageSensor(serial_number, sensor[CONF_NUMBER], sensor[CONF_NAME])
             )
 
-    async_add_entities(entities)
+        async_add_entities(entities)
```

Comment:
Is this right?  `entities` can now be populated with duplicate things if the top-level for loop has more than one monitor configured.

Reply:
Most of the time I think Python's indentation-as-nesting is fine. And then things like this happen. :) Nice catch.